libraries/AP_Proximity/AP_Proximity_Boundary_3D.cpp

```diff
@@ -280,14 +303,58 @@ uint8_t AP_Proximity_Boundary_3D::get_horizontal_object_count() const
     return PROXIMITY_NUM_SECTORS;
 }
 
+// get number of layers
+uint8_t AP_Proximity_Boundary_3D::get_layers() const
+{
+    return PROXIMITY_NUM_LAYERS;
+}
+
 // get an object's angle and distance, used for non-GPS avoidance
 // returns false if no angle or distance could be returned for some reason
 bool AP_Proximity_Boundary_3D::get_horizontal_object_angle_and_distance(uint8_t object_number, float &angle_deg, float &distance) const
 {
     if ((object_number < PROXIMITY_NUM_SECTORS) && _distance_valid[PROXIMITY_MIDDLE_LAYER][object_number]) {
         angle_deg = _angle[PROXIMITY_MIDDLE_LAYER][object_number];
-        distance = _distance[PROXIMITY_MIDDLE_LAYER][object_number];
+        distance = _filtered_distance[PROXIMITY_MIDDLE_LAYER][object_number].get();
         return true;
     }
     return false;
 }
+
+// Return filtered distance for the passed in face
+bool AP_Proximity_Boundary_3D::get_filtered_distance(const Face &face, float &distance) const
+{
+    if (!face.valid()) {
+        return false;
+    }
+
+    if (_distance_valid[face.layer][face.sector]) {
```

Comment:
I'd probably invert the sense of this `if` - you already have one clause returning `false`.

But it doesn't really matter.